### PR TITLE
VCS Team Assignments

### LINKED ISSUE #1422: git (or other vcs) team repos
with team creation the timing will be weird here...  

If all the teams are made (don't necessarily need all members confirmed)...  then calling generate_repos.py with a specific gradeable (note: the repo name doesn't normally have to be the same as a gradeable id, and we want that to continue), but if it is a team repo/team gradeable then we need to know the gradeable_id.  Then we look up the team names in the database.

In order to have access to a team repo, you need to have accepted an invite to that team.

should display full path to team repo on the "manage team" page.

It might be nice if there was a cron job to generate repos for teams, so when a new team was formed, that repo was created automatically (without the instructor or root needing to run the script explicitly).  caution on this though!

### PR DESCRIPTION
**Pull request by James-M-Anderson**
Closes #1422 

Implemented generation of repositories for teams in generate_repos.py.
authentication.py will allow any member of the team, instructor or full access grader to clone the team repository.
The team repository URL is displayed on the team management page.

### CODE REVIEW

File: site/app/views/admin/AdminGradeableView.php [hunk 1]
```diff
@@ -270,7 +270,7 @@ public function show_add_gradeable($type_of_action, AdminGradeable $admin_gradea
                         You are allowed to use the following string replacement variables in format $&#123;&hellip;&#125;<br />
                         <ul style="list-style-position: inside;">
                             <li>gradeable_id</li>
-                            <li>user_id OR repo_id, do not use both</li>
+                            <li>user_id OR team_id OR repo_id, do not use both</li>
```
Comment (MasterOdin) [inline site/app/views/admin/AdminGradeableView.php]: , only use one

File: bin/authentication.py [hunk 2]
```diff
@@ -59,10 +88,20 @@ def check_password(environ, user, password):
         engine, connection, metadata = open_database()
         authenticated = check_database(user, password, connection, metadata)
 
-    if authenticated is not True or user == user_id:
+    if authenticated is not True or (not is_team and user == user_id):
```
Comment (MasterOdin) [inline bin/authentication.py]: If this is a team assignment, is there a case where the "user_id" that's gotten from the path is going to correspond to a user that isn't on the team?

File: bin/generate_repos.py [hunk 3]
```diff
@@ -49,25 +49,59 @@
...
+
+    for team in teams:
+        folder = os.path.join(vcs_course, args.gradeable_id, team.team_id)
+        if not os.path.isdir(folder):
+            os.makedirs(folder, mode=0o770)
+            os.chdir(folder)
+            os.system('git init --bare --shared')
+            for root, dirs, files in os.walk(folder):
+                for entry in files + dirs:
+                    shutil.chown(os.path.join(root, entry), group='www-data')
```
Comment (MasterOdin) [inline bin/generate_repos.py]: Make everything in the if statement be a function:
```
def create_folder(folder)
        if not os.path.isdir(folder):
            os.makedirs(folder, mode=0o770)
            os.chdir(folder)
            os.system('git init --bare --shared')
            for root, dirs, files in os.walk(folder):
                for entry in files + dirs:
                    shutil.chown(os.path.join(root, entry), group='www-data')
```

**Comment by MasterOdin** [comment]
is this ready for review? (I dislike how we're opening the course DB before we're even done with the main submitty DB/basic user authentication).

**Comment by bmcutler** [comment]
We're close.  I'd like to fix the arguments to the gen repo before we commit.  And also I've got a hardcoded URL in there.  And I'll clean up my comments & the <tt>.

Your point about the database is valid too.  If you'd like to fix that I'll do the rest.

File: site/app/views/submission/TeamView.php [hunk 4]
```diff
@@ -81,14 +81,17 @@ public function showTeamPage($gradeable, $teams, $lock) {
...
+	    // Right now, we assume full path is specifified in gradeable
+	    //$this->core->getConfig()->getVcsBaseUrl().$gradeable->getSubdirectory();
 
+	    $repo = $gradeable->getSubdirectory();
 
-	    // TEMPORARY FOR NOW
-	    $repo = "https://submitty-vcs.cs.rpi.edu/git/".$semester.'/'.$course.$gradeable->getSubdirectory();
-
+	    // FIXME: Read from submitty config
+	    $vcs_https_path = "https://submitty-vcs.cs.rpi.edu/git/";
```
Comment (bmcutler) [inline site/app/views/submission/TeamView.php]: @masterodin can you replace this with the system config substitution?

**Comment by bmcutler** [comment]
@MasterOdin I think I'm done with the team repos.  If you can replace that hardcoded http line, and pull the pam check in authenticate earlier, I think we're good to merge.
Thanks

**Pull request merged**